src/lib/mdns/Advertiser_ImplMinimalMdns.cpp

```diff
@@ -0,0 +1,360 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "Advertiser.h"
+
+#include <inttypes.h>
+#include <stdio.h>
+
+#include <mdns/minimal/ResponseSender.h>
+#include <mdns/minimal/Server.h>
+#include <mdns/minimal/responders/IP.h>
+#include <mdns/minimal/responders/Ptr.h>
+#include <mdns/minimal/responders/QueryResponder.h>
+#include <mdns/minimal/responders/Srv.h>
+#include <mdns/minimal/responders/Txt.h>
+#include <support/ReturnMacros.h>
+
+namespace chip {
+namespace Mdns {
+namespace {
+
+template <size_t kSize>
+class StringBuilder
+{
+public:
+    const char * c_str() const { return mBuffer; }
+
+    StringBuilder & Append(const char * s)
+    {
+        size_t available = kSize - mUsed;
+        size_t len       = strlen(s);
+
+        if (len > available)
```

Comment:
`len >= available`?  perhaps also add a test for this class that verifies no buffer overrun with `strlen(s) == kSize`?

Reply:
This was supposed to be debug only. 
However realistically I believe such a 'build debug messages' would be a really nice support class. I will extract it and add tests.

pigweed has stringbuilder however it is C++17 I think (uses string view). Shame.

Reply:
or build this class with `BufferWriter` and reserve the last byte for `\0` ?

Reply:
Created a StringBuilder that uses a buffer writer and includes unit tests.